services/chaospod.go

```diff
@@ -0,0 +1,620 @@
+	ImagePullSecrets              string                                     // Image pull secrets for the chaosPodService.
+	MetricsSink                   metrics.Sink                               // Sink for exporting metrics.
+	CloudServicesProvidersManager cloudservice.CloudServicesProvidersManager // Manager for cloud service providers.
+}
+
+type chaosPodService struct {
+	config ChaosPodServiceConfig
+}
+
+var chaosPodAllowedErrors = map[string]bool{
+	"pod is not running": true,
+	"node is not ready":  true,
+}
+
+// NewChaosPodService create a new chaos pod service instance with the provided configuration.
+func NewChaosPodService(config ChaosPodServiceConfig) (ChaosPodService, error) {
+	if config.Client == nil {
+		return nil, fmt.Errorf("you must provide a non nil Kubernetes client")
+	}
+
+	return &chaosPodService{
+		config: config,
+	}, nil
+}
+
+// CreatePod creates a pod in the Kubernetes cluster.
+func (m *chaosPodService) CreatePod(pod *corev1.Pod) error {
+	return m.config.Client.Create(context.Background(), pod)
+}
+
+// GetChaosPodsOfDisruption retrieves a list of chaos-related pods affected by a disruption event,
+// filtered by the provided labels.
+func (m *chaosPodService) GetChaosPodsOfDisruption(context context.Context, instance *chaosv1beta1.Disruption, ls labels.Set) ([]corev1.Pod, error) {
+	return chaosv1beta1.GetChaosPods(context, m.config.Log, m.config.ChaosNamespace, m.config.Client, instance, ls)
+}
+
+// HandleChaosPodTermination handles the termination of a chaos-related pod during a disruption event.
+func (m *chaosPodService) HandleChaosPodTermination(disruption *chaosv1beta1.Disruption, chaosPod *corev1.Pod) (bool, error) {
+	// Ignore chaos pods that are not being deleted or no longer have the finalizer.
+	if chaosPod.DeletionTimestamp.IsZero() || !controllerutil.ContainsFinalizer(chaosPod, chaostypes.ChaosPodFinalizer) {
+		return false, nil
+	}
+
+	target := chaosPod.Labels[chaostypes.TargetLabel]
+
+	// Check if the target of the disruption is healthy (running and ready).
+	if err := m.config.TargetSelector.TargetIsHealthy(target, m.config.Client, disruption); err != nil {
+		// return the error unless we have a specific reason to ignore it.
+		if !apierrors.IsNotFound(err) && !chaosPodAllowedErrors[strings.ToLower(err.Error())] {
+			return false, err
+		}
+
+		// If the target is not in a good shape, proceed with cleanup phase.
+		m.config.Log.Infow("Target is not likely to be cleaned (either it does not exist anymore or it is not ready), the injector will TRY to clean it but will not take care about any failures", "target", target)
+
+		// Remove the finalizer for the chaos pod since cleanup won't be fully reliable.
+		if err := m.removeFinalizerForChaosPod(chaosPod); err != nil {
+			return false, err
+		}
+
+		return true, nil
+	}
+
+	// If the finalizer cannot be removed yet, return without removing it.
```

Comment:
I'm not 100% sure I understand why we are no longer checking if disruptionHasNoSideEffect (that can hence be removed apparently because no longer used) 

Do you have insights on why it's no longer needed?

Reply:
My bad 🙈 We have to keep this verification 💯

Reply:
It's used here: https://github.com/DataDog/chaos-controller/blob/main/controllers/disruption_controller.go#L794-L797

It allows us to take a fast lane to remove chaos pods we know won't need a specific cleanup of the target without taking care of their exit code.